dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/trackedentity/DefaultTrackedEntityChangeLogService.java

```diff
@@ -0,0 +1,137 @@
+
+  private final TrackerAccessManager trackerAccessManager;
+
+  private final UserService userService;
+
+  private final JdbcTrackedEntityChangeLogStore jdbcTrackedEntityChangeLogStore;
+
+  @Override
+  public Page<TrackedEntityChangeLog> getTrackedEntityChangeLog(
+      UID trackedEntityUid, UID programUid, PageParams pageParams) throws NotFoundException {
+    TrackedEntity trackedEntity =
+        trackedEntityService.getTrackedEntity(trackedEntityUid.getValue());
+    if (trackedEntity == null) {
+      throw new NotFoundException(TrackedEntity.class, trackedEntityUid.getValue());
+    }
+
+    User currentUser = userService.getUserByUsername(CurrentUserUtil.getCurrentUsername());
+    Set<String> trackedEntityAttributes = Collections.emptySet();
+    if (programUid != null) {
+      Program program = validateProgram(programUid.getValue());
+      validateOwnership(currentUser, trackedEntity, program);
+    } else {
+      validateTrackedEntity(trackedEntityUid, currentUser, trackedEntity);
+      trackedEntityAttributes = validateTrackedEntityAttributes(trackedEntity);
+    }
+
+    return jdbcTrackedEntityChangeLogStore.getTrackedEntityChangeLog(
+        trackedEntityUid, trackedEntityAttributes, pageParams);
+  }
+
+  private Program validateProgram(String programUid) throws NotFoundException {
+    Program program = programService.getProgram(programUid);
+    if (program == null) {
+      throw new NotFoundException(Program.class, programUid);
+    }
+
+    return program;
+  }
+
+  private void validateOwnership(User currentUser, TrackedEntity trackedEntity, Program program)
+      throws NotFoundException {
+    if (!trackerAccessManager
+        .canRead(currentUser, trackedEntity, program, currentUser.isSuper())
+        .isEmpty()) {
+      throw new NotFoundException(TrackedEntity.class, trackedEntity.getUid());
+    }
+  }
+
+  private void validateTrackedEntity(
+      UID trackedEntityUid, User currentUser, TrackedEntity trackedEntity)
+      throws NotFoundException {
+    if (!trackerAccessManager.canRead(currentUser, trackedEntity).isEmpty()) {
+      // TODO Do we want to show this error if the user has no access to the org unit? Won't it make
+      // it harder for the end user to understand why the tracked entity is not accessible? What's
+      // the advantage of doing this? Are we giving away any detail by saying the user has no access
+      // to an org unit?
+      throw new NotFoundException(TrackedEntity.class, trackedEntityUid.getValue());
+    }
+  }
+
+  private Set<String> validateTrackedEntityAttributes(TrackedEntity trackedEntity)
+      throws NotFoundException {
+    Set<TrackedEntityAttribute> attributes =
+        trackedEntityAttributeService.getTrackedEntityTypeAttributes(
+            trackedEntity.getTrackedEntityType());
+
+    if (attributes.isEmpty()) {
+      throw new NotFoundException(TrackedEntity.class, trackedEntity.getUid());
```

Comment:
@teleivo does it make sense to throw a `NotFoundException` in this case?
It's gonna make it hard for the end user to figure out the reason why there are no change logs shown, and I think we don't give any hint away by saying that this particular tracked entity has no attributes associated to the tracked entity type. What do you think?